# 1.10. Funciones

## Funciones
- Las funciones permiten la reutilización del código y la simplificación de programas complejos.
- La sintaxis es la siguiente:

```python
def funcname(arg1, arg2,... argN):
    ''' Document String'''
    statements
    return <value>
```
    

- Definimos una función de nombre funcname que acepta los argumentos arg1, arg2,... argN, 
- Está documentada con Document String
- Retorna value, por defecto None.
- Las funciones suelen tener 30 - 50 líneas

In [1]:
def firstfunc():
    ''' No recibe ningún parámetro y devuelve None '''
    print("Hello Jack.")
    print("How are you?")

In [2]:
a = firstfunc()

Hello Jack.
How are you?


In [3]:
a # Hemos guardado None (el valor por defecto)

In [4]:
def firstfunc():
    ''' No recibe ningún parámetro, se limita a devolver el número 5 
    Podríamos devolver una o varias variables con return'''
    return 5

In [5]:
a = firstfunc()

In [6]:
a

5

- Añadimos argumentos.

In [7]:
def firstfunc(username):
    ''' Recibe un parámetro obligatorio, devuelve None '''
    
    print(f"Hello {username}")
    print(f"{username}, how are you?")

In [8]:
name1 = 'Guillermo'

firstfunc(name1)

Hello Guillermo
Guillermo, how are you?


In [9]:
firstfunc() # No le estamos pasando el parámetro obligatorio, por lo que nos devuelve un error

TypeError: firstfunc() missing 1 required positional argument: 'username'

### Return Statement

- Si queremos devolver algún resultado de la función, usamos return.

In [10]:
CTE = 10 # Variable global (por eso están en mayúsulas)

def times(x, y):
    z = x*y*CTE
    return z

- Desde dentro de la función podemos consultar el valor de las variables globales (pero no a la inversa).
- Guardamos en c el valor de z (lo que devuelve la función).
- NO podemos ver z desde fuera de la función.

In [11]:
c = times(4, 5)
print(c)

200


In [12]:
# z no está definido fuera de la función.
# Si intentamos acceder a esta variable nos dará error
z

NameError: name 'z' is not defined

- Del mismo modo, una misma variable puede tener un valor distinto dentro y fuera de la función
- IMPORTANTE: Si cambiamos el valor de la variable x dentro de la función, pero no devolvemos la variable x para que actualice su valor, el valor de esta será el que tenía antes de invocar la función

In [13]:
CTE = 10
x = 20
y = 10

def times(x, y):
    x = x+1 # Cambiamos el valor de x dentro de la función, sumándole 1
    print(x)
    z = x*y*CTE
    return z

solucion = times(x,y)
print(x)

21
20


- Podemos documentar la función:

In [14]:
def times(x, y):
    '''Podemos documentar la función y 
    consultar dicha documentación
    con help()'''
    
    return x*y

In [15]:
c = times(4, 5)
print(c)

20


In [16]:
help(times)

Help on function times in module __main__:

times(x, y)
    Podemos documentar la función y 
    consultar dicha documentación
    con help()



In [17]:
?times

Signature: times(x, y)
Docstring:
Podemos documentar la función y 
consultar dicha documentación
con help()
File:      c:\users\gmelendez\desktop\máster de ia aplicada a bolsa\python-mia-master\module_1\<ipython-input-14-cc6b2a87095a>
Type:      function


- Se pueden retornar múltiples valores como una tupla.
- Puede inducir al error si se intercambian los valores.

In [18]:
eglist = [10, 50, 30, 12, 6, 8, 100]

In [19]:
def eg_func(eglist):
    
    '''Retornamos múltiples variables en una misma función
    En una tupla'''
    
    highest = max(eglist)
    lowest = min(eglist)
    first = eglist[0]
    
    return highest, lowest, first

- Sin asignar da una tupla:

In [20]:
eg_func(eglist)

(100, 6, 10)

- Podemos hacer unpacking:

In [26]:
maximo, minimo, primero = eg_func(eglist)
print(' maximo =', maximo,' minimo =', minimo,' primero =', primero)

 maximo = 100  minimo = 6  primero = 10


In [21]:
maximo, *_ = eg_func(eglist) # El *_ indica que no nos interesa lo que viene después, guardándolo en la varible _

In [22]:
maximo

100

In [23]:
maximo, *_, primero = eg_func(eglist)

In [24]:
print(maximo)
print(primero)

100
10


### Default arguments

- Podemos definir argumentos por defecto de la siguiente forma:

In [25]:
def implicitadd(x, y=3, z=0):
    '''Esta función recibe 3 parámetros
    la x es obligatorio que la indique el usuario
    pero y z tienen valores por defecto'''
    
    print(f"{x} + {y} + {z} = {x+y+z}")
    return x+y+z

In [27]:
implicitadd(10) # Invocamos indicando únicamente el parámetro obligatorio

10 + 3 + 0 = 13


13

In [28]:
implicitadd(10, 1, 2) # Indicando el valor de los tres parámetros (el órden importa)

10 + 1 + 2 = 13


13

In [29]:
implicitadd(10, 1) # Indicamos los valores de x e y

10 + 1 + 0 = 11


11

In [35]:
implicitadd(10, z=1) # Para hacerlo de manera desordenada, hay que indicar el nombre del parámetro al que damos valor

10 + 3 + 1 = 14


14

In [36]:
implicitadd(y=1, z=9, x=2) # Podemos cambiar el órden de todas las variables. Indicar el nombre es la manera correcta de hacer la invocación

2 + 1 + 9 = 12


12

In [37]:
implicitadd(4, 4)
implicitadd(4, 5, 6)
implicitadd(4, z=7)
implicitadd(2, y=1, z=9)
implicitadd(x=1)

4 + 4 + 0 = 8
4 + 5 + 6 = 15
4 + 3 + 7 = 14
2 + 1 + 9 = 12
1 + 3 + 0 = 4


4

### Número no definido de argumentos

- Fuera de Matplotlib no es habitual. Pero que sepáis que esta posibilidad existe.
- Definir una variable como *args almacena todos los argumentos sin clave.
- Definir una variable como **kargs almacena todos los argumentos con clave.

In [30]:
def add_n(first, *args):
    '''Definimos una función que devuelve los elementos, 
    separándolos entre el primero y el resto'''
    
    print(first)
    print(args)

In [31]:
add_n(1)

1
()


In [32]:
add_n(1, 2, 3, 4, 5)

1
(2, 3, 4, 5)


In [33]:
add_n(6.5)

6.5
()


In [34]:
def named_args(**kargs):
    'Imprime los elementos, uniéndolos por clave:valor'
    
    print(kargs)

In [35]:
named_args(x=3, animal='mouse', z=(1+2j))

{'x': 3, 'animal': 'mouse', 'z': (1+2j)}


In [36]:
def all_args(first, *args, x=1, **kargs):
    
    'Podemos complicarlo tanto como queramos'
    
    print(first)
    print(args)
    print(x)
    print(kargs)

In [37]:
all_args(1, 2, 3, x=3, animal='mouse', z=(1+2j))

1
(2, 3)
3
{'animal': 'mouse', 'z': (1+2j)}


###  Variables Global y Local 

- Las variables definidas dentro de la función son locales.
- Las variables definidas fuera de la función son globales.

In [38]:
eg1 = [1, 2, 3, 4, 5]

In [39]:
def egfunc1():
    x = 1
    print(eg1)
    
egfunc1()

[1, 2, 3, 4, 5]


In [40]:
# eg1 es una variable global
# Pero x es una variable local. No puedo acceder a ella desde fuera de la función
x

20

In [41]:
x = 5

def egfunc1():
    x = 1
    print(x)
    print(eg1)
    
egfunc1()

print(x) # A menos que la función devuelva x y lo reasignemos, el valor de x no ha sido alterado una vez termine la función

1
[1, 2, 3, 4, 5]
5


In [50]:
x = [5]

def egfunc1():
    x.append(1)
    print(x)
    
egfunc1()

print(x) # Sin embargo, a la hora de modificar estructuras más complejas (listas, tuplas, diccionarios), estas sí se actualizan dentro de una función (porque son una referencia a su posición en la memoria)

[5, 1]
[5, 1]


In [42]:
def egfunc1():
    ''' Podemos definir una función dentro de una función...'''
    
    x=1
    
    def thirdfunc():
        x=2
        print("Inside thirdfunc x =", x)
        
    thirdfunc()
    
    print("Inside egfunc1 x =", x)

In [43]:
x = "Hola"

egfunc1()

print("Outside x =", x)

Inside thirdfunc x = 2
Inside egfunc1 x = 1
Outside x = Hola


- Las variables **global**  hacen a las funciones difíciles de reusar y deben de ser utilizadas poco frecuentemente.

In [44]:
eg3 = [1, 2, 3, 4, 5]

In [45]:
def egfunc1():
    x = 1.0 # variable local de egfunc1
    
    def thirdfunc():
        global x # Definimos una variable como global 
        x = 2.0
        print("Inside thirdfunc x =", x) 
    thirdfunc()
    
    print("Inside egfunc1 x =", x)

In [46]:
egfunc1()
print("Globally defined x =",x)

Inside thirdfunc x = 2.0
Inside egfunc1 x = 1.0
Globally defined x = 2.0


### Lambda Functions

- Son funciones definidas en una única línea.
- Suelen usarse como inputs de otras funciones
- Definidas con la palabra **lambda**.

In [47]:
def square(x):
    '''Típica función que eleva el input al cuadrado'''
    return x*x

In [48]:
square(2)

4

In [49]:
# Misma función a través de Lambda 
# Nombre que queremos darle a la función, lambda, nombre del parámetro que recibimos, lo que hace la función (retorno)

square = lambda x: x*x

In [50]:
square(2)

4

- Utiles por ejemplo para ordenar listas.

In [51]:
lista_odenar = ['casa', 'hola', 'u', 'bbbbbbb', 'zzzzzzzzzzz']

In [52]:
lista_odenar.sort(key = lambda x: x[0]) # Ordenamos por la primera letra

In [53]:
lista_odenar

['bbbbbbb', 'casa', 'hola', 'u', 'zzzzzzzzzzz']

In [54]:
lista_odenar.sort(key = lambda x: len(x)) # Ordenamos por el tamaño de los elementos

In [55]:
lista_odenar

['u', 'casa', 'hola', 'bbbbbbb', 'zzzzzzzzzzz']

### Composición de funciones
- Las funciones se pueden pasar como parámetros a otras funciones.

In [56]:
def double(x):
    return 2*x

def square(x):
    return x*x

In [57]:
double(square(3))

18

___
# Ejercicios

**1.10.1.** Escribe una función a la que se le pase una vector de números y calcule su media.

**1.10.2.** Escribe una función que calcule el IVA (21%) de un producto dado su precio de venta sin IVA y devuelva el precio total.

**1.10.3.** Escribe una función que reconozca palíndromos. Un palíndromo es una palabra que se lee igual al derecho que al revés.

**1.10.4.** Escribe una función Python que a partir de una cantidad en euros, y del tipo de cambio del día, retorne el equivalente en libras teniendo en cuenta que la casa de cambio retiene una comisión del 2% sobre el total de la operación. Nota: EUR/GBP = 0.8624

**1.10.5.** Escribe una función que calcule el área de un círculo recibiendo como parámetro su radio (valor por defecto 1). El módulo math de Python contiene una constante para PI.


**1.10.6.** Escribe una función que devuelva una lista con n números aleatorios entre 0 y 10 generados con una distribución uniforme (n es un parámetro obligatorio). El módulo random de Python contiene una función uniform.